src/main.cpp

```diff
@@ -1142,6 +1142,13 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
         }
     }
 
+    // checking txs commitment tree validity
+    SidechainTxsCommitmentBuilder scCommitmentBuilder;
+    if (!scCommitmentBuilder.add(tx)) {
+        return state.DoS(100, error("%s(): SidechainTxsCommitmentBuilder failed\n", __func__),
+                        CValidationState::Code::INVALID, "bad-txns-txscommitmenttree-fail");
+    }
+
```

Comment:
I see here that the check prevents a single transaction to exceed the capacity of the commitment tree, and that in case the sender node is banned, it's a very important check.

I wonder if we could find another place for this, for instance at the end of `Sidechain::checkTxSemanticValidity()`, so that it is performed both for the mempool and the block validation.

Reply:
Checks on the commitment tree are performed in different places, depending on which entity we are currently checking, covering all the possibilities:
- for incoming blocks, we test the correctness of the txs commitment `CheckBlock()`, by building a `SidechainTxsCommitmentBuilder` and progressively adding all the txs an certs contained within
- we do the same thing in `CreateNewBlock()` for blocks we are generating i.e. through `getblocktemplate()`
- for an incoming tx to be added to the mempool, we do this check in `AcceptTxToMemoryPool()->CheckTransaction()->CheckTransactionWithoutProofVerification()`
- for an incoming cert to be added to the mempool, we do this check directly in `AcceptCertificateToMemoryPool()`, as for this check we need a valid view to extract the sidechain from `CSidechainMap`, given its sidechain id; so although this check being theoretically pure semantic, we cannot move it to `Sidechain::checkTxSemanticValidity()`